File: android/src/main/kotlin/pers/hal42/util/StopWatch.java
/**
 * wrap common usages of System.currentTimeMillis
 */

package pers.hal42.util;

public class StopWatch {
  long started;
  long stopped;
  boolean running;

  public boolean isRunning() {
    return running; //is atomic so no synch needed.
  }

  public long startedAt() {
    return started;
  }

  public double seconds() { //can be read while running
    return ((double) millis()) / Ticks.perSecond;
  }

  public long millis() { //can be read while running
    return (running ? System.currentTimeMillis() : stopped) - started;
  }

  public void Start() {
    stopped = started = System.currentTimeMillis();
    running = true;
  }

  public long Stop() {
    if(running) {
      stopped = System.currentTimeMillis();
      running = false;
    }
    return millis();
  }

  public void Reset() {
    running = false;
    stopped = started = 0;
  }

  public StopWatch(boolean hitTheFloorRunning) {
    Reset();
    if(hitTheFloorRunning) {
      Start();
    }
  }

  public StopWatch() {
    this(true);
  }
}

File: android/src/main/kotlin/pers/hal42/util/StopWatch.kt
/**
 * wrap common usages of System.currentTimeMillis
 */

package pers.hal42.util

class StopWatch @JvmOverloads constructor(hitTheFloorRunning: Boolean = true) {
  internal var started: Long = 0
  internal var stopped: Long = 0
  //is atomic so no synch needed.
  var isRunning: Boolean = false
    internal set

  fun startedAt(): Long {
    return started
  }

  fun seconds(): Double { //can be read while running
    return millis().toDouble() / Ticks.perSecond
  }

  fun millis(): Long { //can be read while running
    return (if (isRunning) System.currentTimeMillis() else stopped) - started
  }

  fun Start() {
    started = System.currentTimeMillis()
    stopped = started
    isRunning = true
  }

  fun Stop(): Long {
    if (isRunning) {
      stopped = System.currentTimeMillis()
      isRunning = false
    }
    return millis()
  }

  fun Reset() {
    isRunning = false
    started = 0
    stopped = started
  }

  init {
    Reset()
    if (hitTheFloorRunning) {
      Start()
    }
  }
}
